### PR TITLE
Add hyphen to allowed characters in stream-name

### PR DESCRIPTION
**Pull request by james-smith-za**
CAM would like to be able to call CBF streams things like `antenna-channelised-voltage`, rather than
`antenna_channelised_voltage`. I think this commit should enable it.

(Possibly) closes NGC-914.

### CODE REVIEW

**Comment by james-smith-za** [comment]
Ok, Jenkins fails with an ominous "Failed to process config: Invalid \\escape: line 8 column 37 (char 174)".
So clearly it doesn't like my escape of the hyphen for some reason.

**Comment by ludwigschwardt** [comment]
Maybe add a `r` in front of the string, since the backslash is meant for `re` instead?

**Comment by james-smith-za** [comment]
> Maybe add a `r` in front of the string, since the backslash is meant for `re` instead?
> 
> In other words: `r"^[A-Za-z0-9_\-]+$"`.
> 
> Oops sorry, scratch that - it's not Python but JSON :-)

Unfortunately I didn't see your edit before I pushed a change. I give up for the afternoon though.

File: src/katsdpcontroller/product_config.py [hunk 1]
```diff
@@ -2074,7 +2074,7 @@ def _upgrade(config):
         config["version"] = "3.0"
 
     # Upgrade to latest 3.x
-    config["version"] = "3.3"
+    config["version"] = str(schemas.PRODUCT_CONFIG.versions[-1])
```
Comment (bmerry) [inline src/katsdpcontroller/product_config.py:2077]: I remember this is what I suggested, but now that I see it I wonder if
```suggestion
    config["version"] = str(max(schemas.PRODUCT_CONFIG.versions))
```
captures the intent better.

**Review by bmerry — APPROVED** [review]
Looks good. One bit of bikeshedding which you can feel free to apply or not.

**Comment by bmerry** [comment]
Another bit of optional bikeshedding: if you move the `-` to the end of the character class instead of the beginning, it's clearer that both `_` and `-` are available.

File: src/katsdpcontroller/product_config.py (hunk 1 above)
Comment (james-smith-za) [reply src/katsdpcontroller/product_config.py:2077]: Does this interpret the strings properly as float numbers? I'm a little hazy on how the parsing works from j2 template to json to python.

Comment (bmerry) [reply src/katsdpcontroller/product_config.py:2077]: schemas.PRODUCT_CONFIG.versions are instances of `packaging.version.Version`, which sort the way versions should (see `katsdpcontroller.schemas._normalise_version`).

File: src/katsdpcontroller/schemas/product_config.json.j2 [hunk 2]
```diff
@@ -1,15 +1,16 @@
 {% macro versions() %}
-["2.4", "2.5", "2.6", "3.0", "3.1", "3.2", "3.3"]
+["2.4", "2.5", "2.6", "3.0", "3.1", "3.2", "3.3", "3.4"]
 {% endmacro %}
 
 {% macro validate(version) %}
 {% set input_additional = "false" if version >= "3.0" else "true" %}
+{% set stream_name_pattern = "^[-A-Za-z0-9_]+$" if version >= "3.4" else "^[A-Za-z0-9_]+$" %}
```
Comment (ludwigschwardt) [inline src/katsdpcontroller/schemas/product_config.json.j2:7]: Thanks, now I've (re-)learnt something about re. 😄

**Review by ludwigschwardt — APPROVED** [review]

**Pull request merged**